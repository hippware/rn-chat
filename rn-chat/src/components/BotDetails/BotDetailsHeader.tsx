import React from 'react'
import {View, Animated, Alert, Image, TouchableOpacity, StyleSheet, Clipboard} from 'react-native'
import {observer, inject} from 'mobx-react/native'
import {k, width, height} from '../Global'
import {colors} from '../../constants'
import {getSnapshot, isAlive} from 'mobx-state-tree'
import BotButtons from './BotButtons'
import UserInfoRow from './UserInfoRow'
import {RText} from '../common'
import BotDetailsMap from '../map/BotDetailsMap'
import {Actions} from 'react-native-router-flux'
import {IBot} from 'wocky-client'

type Props = {
  bot: IBot
  notificationStore?: any // TODO proper type
  analytics?: any // TODO proper type
  scale: number
}

type State = {
  fadeAnim: any
}

const DOUBLE_PRESS_DELAY = 300

@inject('notificationStore', 'analytics')
@observer
class BotDetailsHeader extends React.Component<Props, State> {
  lastImagePress?: number
  userInfo: any

  constructor(props: Props) {
    super(props)
    this.state = {
      fadeAnim: new Animated.Value(0),
    }
  }

  handleImagePress = () => {
    const now = new Date().getTime()

    if (this.lastImagePress && now - this.lastImagePress < DOUBLE_PRESS_DELAY) {
      delete this.lastImagePress
      this.handleImageDoublePress()
    } else {
      this.lastImagePress = now
    }
  }

  handleImageDoublePress = () => {
    if (!this.props.bot.isSubscribed) {
      this.subscribe()
    }
  }

  unsubscribe = () => {
    Alert.alert('', 'Are you sure you want to remove this from your saved bots?', [
      {text: 'Cancel', style: 'cancel'},
      {
        text: 'Remove',
        style: 'destructive',
        onPress: () => this.props.bot.unsubscribe(),
      },
    ])
  }

  subscribe = () => {
    this.props.bot.subscribe()
    this.setState({fadeAnim: new Animated.Value(1)})
    setTimeout(() => {
      Animated.timing(this.state.fadeAnim, {toValue: 0}).start()
    }, 500)
    this.props.analytics.track('bot_save', getSnapshot(this.props.bot))
  }

  copyAddress = () => {
    Clipboard.setString(this.props.bot.address)
    this.props.notificationStore.flash('Address copied to clipboard üëç')
  }

  render() {
    const {bot, scale} = this.props
    if (!bot || !isAlive(bot)) return null
    return (
      <View style={{flex: 1}}>
        <View
          style={{
            height: scale === 0 ? height - 60 * k : width,
            backgroundColor: 'white',
            overflow: 'hidden',
          }}
        >
          <BotDetailsMap
            bot={bot}
            onMapPress={() => Actions.refresh({scale: 0})}
            onImagePress={() => {
              Actions.refresh({scale: scale === 1 ? 0.5 : scale + 0.5})
            }}
            scale={scale}
          />
          <Animated.View
            pointerEvents="none"
            style={[{opacity: this.state.fadeAnim}, styles.botAddedContainer]}
          >
            <Image source={require('../../../images/iconBotAdded.png')} />
          </Animated.View>
        </View>
        {scale > 0 && (
          <View>
            <View style={styles.rowContainer}>
              <BotButtons
                bot={bot}
                subscribe={this.subscribe}
                unsubscribe={this.unsubscribe}
                isSubscribed={bot.isSubscribed}
                copyAddress={this.copyAddress}
              />
              <UserInfoRow bot={bot} copyAddress={this.copyAddress} />
              {!!bot.description && (
                <View style={styles.descriptionContainer}>
                  <RText numberOfLines={0} size={16} weight="Light" color={colors.DARK_PURPLE}>
                    {bot.description}
                  </RText>
                </View>
              )}
              {bot.guest && <GeofenceCTA bot={bot} />}
            </View>
            <View
              style={[
                styles.rowContainer,
                {
                  height: 45,
                  alignSelf: 'stretch',
                  marginTop: 8.5,
                  flexDirection: 'row',
                  alignItems: 'center',
                },
              ]}
            >
              <Image
                style={{width: 14, height: 14}}
                source={require('../../../images/postsIcon.png')}
              />
              <RText
                size={15}
                color={colors.DARK_PURPLE}
                style={{marginLeft: 7, letterSpacing: 0.3}}
              >
                Posts
              </RText>
              <RText size={12} color={colors.DARK_GREY} style={{marginLeft: 7}}>
                {bot.totalItems}
              </RText>
            </View>
            <View style={{height: 1, width}} />
          </View>
        )}
      </View>
    )
  }
}

const GeofenceCTA = observer(({bot}) => (
  <View style={{flexDirection: 'row'}}>
    <Image source={require('../../../images/footOpaquePink.png')} style={{width: 26, height: 34}} />
    <TouchableOpacity onPress={() => Actions.visitors({item: bot.id})}>
      <View style={{marginLeft: 10 * k}}>
        <RText color={colors.PINK} size={15}>
          See Who's Here
        </RText>
        <RText color={colors.DARK_GREY} size={12}>
          {bot.visitorsSize} people
        </RText>
      </View>
    </TouchableOpacity>
  </View>
))

export default BotDetailsHeader

const styles = StyleSheet.create({
  rowContainer: {
    backgroundColor: 'white',
    padding: 20 * k,
  },
  descriptionContainer: {
    paddingBottom: 15 * k,
  },
  botAddedContainer: {
    height: width,
    position: 'absolute',
    left: 0,
    right: 0,
    justifyContent: 'center',
    alignItems: 'center',
  },
})
