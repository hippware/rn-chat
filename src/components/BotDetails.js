import React from 'react'
import {
    View,
    TouchableWithoutFeedback,
    Text,
    ScrollView,
    Animated,
    Alert,
    TouchableOpacity,
    Image
} from 'react-native'
import { Actions } from 'react-native-router-native'
import Screen from './Screen'
import botFactory from '../factory/botFactory'
import Map from './Map'
import { Annotation } from 'react-native-mapbox-gl'
import GradientHeader from './GradientHeader'
import { k, width, height, defaultCover } from './Global'
import BotAvatar from './BotAvatar'
import Avatar from './Avatar'
import { observer } from 'mobx-react/native'
import { observable } from 'mobx'
import botStore from '../store/botStore'
import location from '../store/locationStore'
import Bot, { VISIBILITY_PUBLIC, VISIBILITY_OWNER, LOCATION, NOTE, IMAGE } from '../model/Bot'
import autobind from 'autobind-decorator'
import statem from '../../gen/state'
import PhotoGrid from './PhotoGrid'
import model from '../model/model'
import { when } from 'mobx'
import BotNavBar from './BotNavBar'

const DOUBLE_PRESS_DELAY = 300
function Header (props) {
    return <View style={{
        backgroundColor: 'rgba(255,255,255,0.85)',
        flexDirection: 'row',
        height: 41 * k,
        shadowOffset: {height: 1, width: 0},
        shadowRadius: 2,
        shadowOpacity: 0.12,
    }}>
        <View style={{flex: 1, justifyContent: 'center'}}>
            <Text numberOfLines={1} style={{
                paddingLeft: 26 * k,
                fontSize: 14 * k,
                color: 'rgb(63,50,77)',
                fontFamily: 'Roboto-Regular'
            }}>{props.children}</Text>
        </View>
    </View>
}

@autobind
@observer
export default class extends React.Component {

    constructor (props) {
        super(props)
        this.loading = false
        this.state = {
            top: new Animated.Value(this.props.fullMap ? height : 0),
            fullMap: !!this.props.fullMap,
            fadeAnim: new Animated.Value(0),
            showNavBar: true,
            navBarHeight: new Animated.Value(70),
        }
    }

    async loadMoreImages () {
        if (botStore.bot && botStore.bot.imagesCount && botStore.bot.images.length && botStore.bot.imagesCount > botStore.bot.images.length) {
            if (!this.loading) {
                this.loading = true
                console.log('LOAD MORE IMAGES', botStore.bot.images[botStore.bot.images.length - 1].item)
                await botStore.loadImages(botStore.bot.images[botStore.bot.images.length - 1].item)
                this.loading = false
            }
        }
    }

    onScrollStart () {
        // display 'no more images'
        if (botStore.bot.imagesCount > 0 && botStore.bot.imagesCount === botStore.bot.images.length) {
            this.setState({showNoMoreImages: true})
        }
    }

    onScrollEnd (event) {
        // load more images
        if (!this.state.showNavBar) {
            console.log('SCROLL END!', botStore.bot.imagesCount, botStore.bot.images.length)
            // this.setState({showNavBar: true})
            // Animated.timing(
            //   this.state.navBarHeight,
            //   {toValue: 70}
            // ).start();
        }
        this.setState({showNoMoreImages: false})

    }

    onScroll (event) {
        if (event.nativeEvent.contentOffset.y + height + 200 >= event.nativeEvent.contentSize.height) {
            this.loadMoreImages()
        }
        // if (this.state.showNavBar){
        //   this.setState({showNavBar: false})
        //   Animated.timing(
        //     this.state.navBarHeight,
        //     {toValue: 0}
        //   ).start();
        // }
        // switch nav bar is scroll position is below threshold
        // const y = event.nativeEvent.contentOffset.y;
        // const limit = 50 * k;
        // const upper = 240*k;
        // if (y > limit) {
        //   Animated.timing(          // Uses easing functions
        //     this.state.fadeAnim,    // The value to drive
        //     {toValue: Math.max(0, 1 - y/upper)}            // Configuration
        //   ).start();
        //   if (!this.state.showNavBar) {
        //     this.setState({showNavBar: true});
        //   }
        // } else {
        //   Animated.timing(          // Uses easing functions
        //     this.state.fadeAnim,    // The value to drive
        //     {toValue: Math.min(1, 1 - y/upper)}            // Configuration
        //   ).start();
        //   if (this.state.showNavBar) {
        //     this.setState({showNavBar: false});
        //   }
        // }
    }

    async componentWillMount () {
        if (this.props.item) {
            botStore.bot = botFactory.create({id: this.props.item})
            when(() => model.connected, botStore.load)
        }
    }

    hideFullMap () {
        this.setState({fullMap: true})
        Animated.timing(
            this.state.top,
            {toValue: 0}
        ).start()
        Actions.refresh({})
    }

    showFullMap () {
        this.setState({fullMap: true})
        Animated.timing(
            this.state.top,
            {toValue: height}
        ).start()
        Actions.refresh({leftButton: {icon: require('../../images/iconClose.png'), onPress: this.hideFullMap}})
    }

    onLayout (event) {
        var layout = event.nativeEvent.layout
        this.setState({currentScreenWidth: layout.width, currentScreenHeight: layout.height})
    }

    unsubscribe () {
        Alert.alert(null, 'Are you sure you want to unsubscribe?', [
            {text: 'Cancel', style: 'cancel'},
            {text: 'Unsubscribe', style: 'destructive', onPress: () => botStore.unsubscribe()}
        ])
    }

    subscribe () {
        botStore.subscribe()
        // do animation
        this.setState({fadeAnim: new Animated.Value(1)})
        setTimeout(() => {
            Animated.timing(
                this.state.fadeAnim,
                {toValue: 0}
            ).start()

        }, 500)
    }

// ...

    /**
     * Double Press recognition
     * @param  {Event} e
     */
    handleImagePress (e) {
        const now = new Date().getTime()

        if (this.lastImagePress && (now - this.lastImagePress) < DOUBLE_PRESS_DELAY) {
            delete this.lastImagePress
            this.handleImageDoublePress(e)
        }
        else {
            this.lastImagePress = now
        }
    }

    handleImageDoublePress (e) {
        const bot = botStore.bot
        if (!bot.isSubscribed) {
            this.subscribe()
        }
    }

    render () {
        const bot = botStore.bot
        if (!bot) {
            console.log('ERROR: No bot defined', this.props.item)
            return <Screen/>
        }
        const isDay = location.isDay
        const isOwn = !bot.owner || bot.owner.isOwn
        const coef = bot.image && bot.image.width ? (width - 34 * k) / bot.image.width : 0
        const profile = bot.owner
        if (!profile) {
            console.log('ERROR: NO BOT PROFILE!')
            return <Screen/>
        }
        const source = bot.image && bot.image.source
        return <View style={{flex: 1, backgroundColor: location.isDay ? 'white' : 'rgba(49,37,62,0.90)'}}>
            <ScrollView style={{paddingTop: 70 * k}} onScrollEndDrag={this.onScrollEnd}
                        onScrollBeginDrag={this.onScrollStart} onScroll={this.onScroll} scrollEventThrottle={1}>
                <View style={{width: 375 * k, height: 275 * k}}>
                    <TouchableWithoutFeedback onPress={this.handleImagePress}>
                        <Image style={{width: 375 * k, height: 275 * k}}
                               source={source || defaultCover[bot.coverColor % 4]}/>
                    </TouchableWithoutFeedback>
                    {isOwn && <TouchableOpacity onPress={() => statem.logged.botEdit({item: bot.id})}
                                                style={{
                                                    borderRadius: 2,
                                                    backgroundColor: 'rgba(255,255,255,0.75)',
                                                    position: 'absolute',
                                                    justifyContent: 'center',
                                                    alignItems: 'center',
                                                    bottom: 20 * k,
                                                    right: 20 * k,
                                                    height: 30 * k,
                                                    width: 36 * k
                                                }}>
                        <Image source={require('../../images/iconEditBot.png')}/>
                    </TouchableOpacity>}
                    <Animated.View pointerEvents="none" style={{
                        width: 375 * k, height: 275 * k, opacity: this.state.fadeAnim,
                        position: 'absolute', justifyContent: 'center', alignItems: 'center'
                    }}>
                        <Image source={require('../../images/iconBotAdded.png')}/>
                    </Animated.View>
                </View>
                <View style={{paddingTop: 15 * k, paddingLeft: 20 * k, paddingRight: 20 * k}}>
                    {!isOwn && !bot.isSubscribed && <TouchableOpacity onPress={this.subscribe} style={{
                        height: 40 * k,
                        borderWidth: 0,
                        backgroundColor: 'rgb(254,92,108)',
                        borderRadius: 2 * k,
                        justifyContent: 'center',
                        alignItems: 'center'
                    }}>
                        <Text
                            style={{fontSize: 11 * k, letterSpacing: 0.5, fontFamily: 'Roboto-Medium', color: 'white'}}>ADD
                            BOT</Text>
                    </TouchableOpacity>}
                    {!isOwn && !!bot.isSubscribed && <TouchableOpacity onPress={this.unsubscribe} style={{
                        height: 40 * k,
                        flexDirection: 'row',
                        justifyContent: 'center',
                        alignItems: 'center',
                        borderWidth: 0,
                        backgroundColor: 'rgb(228,228,228)',
                        borderRadius: 2 * k
                    }}>
                        <View style={{padding: 10 * k}}><Image source={require('../../images/iconCheckBotAdded.png')}/></View>
                        <Text style={{
                            fontSize: 11 * k,
                            letterSpacing: 0.5,
                            fontFamily: 'Roboto-Medium',
                            color: 'rgb(99,62,90)'
                        }}>BOT ADDED</Text>
                    </TouchableOpacity>}
                </View>
                <View style={{
                    paddingTop: 15 * k,
                    paddingBottom: 15 * k,
                    paddingLeft: 20 * k,
                    paddingRight: 20 * k,
                    flexDirection: 'row',
                    alignItems: 'center'
                }}>
                    <View style={{paddingRight: 11 * k}}><Avatar size={36} profile={profile}
                                                                 source={profile.avatar && profile.avatar.source}
                                                                 title={profile.displayName} isDay={location.isDay}
                                                                 disableStatus borderWidth={0}/></View>
                    <View style={{flex: 1}}><Text style={{
                        fontFamily: 'Roboto-Italic',
                        fontSize: 13,
                        letterSpacing: -0.1,
                        color: 'rgb(114,100,109)'
                    }}>@{profile.handle}</Text></View>
                    {location.location && bot.location && <View>
                        <Image source={require('../../images/buttonViewMapBG.png')}/>
                        <TouchableOpacity onPress={statem.botDetails.map} style={{
                            position: 'absolute',
                            top: 0,
                            bottom: 0,
                            right: 0,
                            left: 0,
                            justifyContent: 'center',
                            alignItems: 'center',
                            flexDirection: 'row',
                            backgroundColor: 'transparent'
                        }}>
                            <View style={{paddingRight: 5}}><Image
                                source={require('../../images/iconBotLocation.png')}/></View>
                            <Text style={{fontFamily: 'Roboto-Regular', fontSize: 13, color: 'rgb(63,50,77)'}}>
                                {location.distanceToString(
                                    location.distance(location.location.latitude, location.location.longitude, bot.location.latitude, bot.location.longitude)
                                )}
                            </Text>
                        </TouchableOpacity>
                    </View>}
                </View>
                {!!bot.description && <View style={{paddingLeft: 20 * k, paddingRight: 20 * k, paddingBottom: 15 * k}}>
                    <Text numberOfLines={0} style={{
                        fontFamily: 'Roboto-Light',
                        fontSize: 15,
                        color: location.isDay ? 'rgb(63,50,77)' : 'white'
                    }}>{bot.description}</Text>
                </View>
                }
                {!isOwn && !bot.imagesCount && <View style={{
                    height: 201 * k,
                    backgroundColor: 'rgb(242,243,245)',
                    justifyContent: 'center',
                    alignItems: 'center'
                }}>
                    <Image source={require('../../images/attachPhotoGray.png')}/>
                    <Text style={{fontFamily: 'Roboto-Regular', fontSize: 15, color: 'rgb(186,186,186)'}}>No photos
                        added</Text>
                </View>}
                <PhotoGrid isOwn={isOwn} images={bot.images} onAdd={statem.botDetails.addPhoto}
                           onView={index => statem.botDetails.editPhotos({index})}/>
                {this.state.showNoMoreImages &&
                <View style={{paddingTop: 10, alignItems: 'center', paddingBottom: 21}}><Image
                    source={require('../../images/graphicEndPhotos.png')}/></View>}
            </ScrollView>
            {this.state.showNavBar && <BotNavBar bot={bot}/>}
        </View>
    }
}