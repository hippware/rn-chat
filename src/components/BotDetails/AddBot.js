// @flow

import React from 'react';
import {Text, TouchableOpacity, StyleSheet, Image} from 'react-native';
import {colors} from '../../constants';
import {k} from '../../globals';
import {observer} from 'mobx-react/native';

export default observer(({subscribe, unsubscribe, isSubscribed}) => {
  let onPress, buttonStyle, image, text, textStyle;
  if (isSubscribed) {
    onPress = unsubscribe;
    buttonStyle = [styles.addBotButton, {backgroundColor: colors.WHITE, borderColor: colors.GREY}];
    image = require('../../../images/iconCheckBotAdded.png');
    text = 'SAVED';
    textStyle = [styles.addBotText, {color: colors.PURPLE}];
  } else {
    onPress = subscribe;
    buttonStyle = styles.addBotButton;
    image = require('../../../images/saveIcon.png');
    text = 'SAVE';
    textStyle = styles.addBotText;
  }
  return (
    <TouchableOpacity onPress={onPress} style={buttonStyle}>
      <Image source={image} style={styles.addBotIcon} resizeMode='contain' />
      <Text style={textStyle}>
        {text}
      </Text>
    </TouchableOpacity>
  );
});

const styles = StyleSheet.create({
  addBotButton: {
    flexDirection: 'row',
    height: 40 * k,
    marginTop: -20 * k,
    width: 150 * k,
    alignSelf: 'center',
    backgroundColor: colors.PINK,
    borderRadius: 5 * k,
    borderColor: colors.PINK,
    borderWidth: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
  addBotText: {
    fontSize: 15,
    letterSpacing: 0.5,
    fontFamily: 'Roboto-Medium',
    color: 'white',
    marginLeft: 5 * k,
  },
});
