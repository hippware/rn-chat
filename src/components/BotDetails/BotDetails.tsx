import React from 'react'
import {View, Clipboard, TouchableOpacity} from 'react-native'
import {observable, runInAction} from 'mobx'
import {observer, inject} from 'mobx-react/native'
import {k} from '../Global'
import {colors} from '../../constants'
import {IProfile, IBot, IWocky} from 'wocky-client'
import BotPostCard from './BotPostCard'
import {RText, Spinner} from '../common'
import AddBotPost from './AddBotPost'
import Header from './BotDetailsHeader'
import {isAlive} from 'mobx-state-tree'
import Separator from './Separator'
import DraggablePopupList from '../common/DraggablePopupList'
import {Actions} from 'react-native-router-flux'
import {navBarStyle} from '../styles'
import NotificationStore from '../../store/NotificationStore'

type Props = {
  botId: string
  isNew?: boolean
  params?: any
  wocky?: IWocky
  analytics?: any
  notificationStore?: NotificationStore
  homeStore?: any
  navigation: any
  isActive: boolean
}

@inject('wocky', 'analytics', 'notificationStore', 'homeStore')
@observer
export default class BotDetails extends React.Component<Props> {
  @observable bot?: IBot
  @observable owner?: IProfile
  @observable numToRender: number = 8
  list: any
  viewTimeout: any

  static navigationOptions = ({navigation}) => {
    const {isNew, bot, notificationStore} = navigation.state.params
    const backAction = isNew ? () => Actions.popTo('home') : Actions.pop
    return {
      backAction,
      fadeNavConfig: {
        back: true,
        backAction,
        title: bot && (
          <NavTitle
            bot={bot}
            onLongPress={() => {
              Clipboard.setString(bot.address)
              notificationStore.flash('Address copied to clipboard 👍')
            }}
          />
        ),
      },
    }
  }

  _footerComponent: any = observer(() => {
    if (!this.bot) return null

    if (this.props.wocky!.connected && this.bot && isAlive(this.bot) && this.bot.posts.loading)
      return <Loader />

    return <View style={{backgroundColor: 'white', height: 100 * k}} />
  })

  async componentWillMount() {
    const {wocky, analytics, botId, homeStore, navigation} = this.props
    runInAction(() => {
      this.bot = wocky!.getBot({id: botId})
    })
    if (!this.bot) {
      return
    }

    // send all injected props + bot "up" to static context
    this.props.navigation.setParams({...this.props, bot: this.bot})

    homeStore.select(this.bot.id)
    homeStore.setFocusedLocation(this.bot.location)
    await wocky!.loadBot(botId)

    this.viewTimeout = setTimeout(() => {
      if (this.bot && isAlive(this.bot))
        analytics.track('bot_view', {id: this.bot.id, title: this.bot.title})
    }, 7000)

    // deep-linking to visitors
    if (navigation.state.params.params === 'visitors') {
      // todo: why doesn't Botdetails pop back down when nav'ing to visitors?
      Actions.visitors({botId})
    }
  }

  componentWillUnmount() {
    if (this.viewTimeout) {
      clearTimeout(this.viewTimeout)
    }
  }

  scrollToNewestPost = () => {
    this.list.wrappedInstance.scrollToIndex({
      index: 0,
      viewPosition: 0.5,
    })
  }

  renderItem = ({item}) => <BotPostCard item={item} bot={this.bot!} />

  renderSeparator = () => (
    <View style={{backgroundColor: 'white'}}>
      <Separator />
    </View>
  )

  render() {
    const {bot} = this
    if (!bot || !isAlive(bot)) {
      return null
    }

    return (
      <View pointerEvents="box-none" style={{flex: 1}}>
        <DraggablePopupList
          isActive={this.props.isActive}
          data={!bot.error && bot.isSubscribed ? bot.posts.list.slice() : []}
          ref={r => (this.list = r)}
          contentContainerStyle={{
            flexGrow: 1,
          }}
          ListFooterComponent={this._footerComponent}
          initialNumToRender={this.numToRender}
          headerInner={<Header bot={this.bot!} {...this.props} />}
          ItemSeparatorComponent={this.renderSeparator}
          renderItem={this.renderItem}
          keyExtractor={item => item.id}
          bounces={false}
          keyboardDismissMode="on-drag"
        />
        {!bot.error && bot.isSubscribed && (
          <AddBotPost bot={bot} afterPostSent={this.scrollToNewestPost} />
        )}
      </View>
    )
  }
}

const NavTitle = ({bot, onLongPress}) => {
  const {titleStyle} = navBarStyle
  if (!bot) {
    return null
  }
  return (
    <TouchableOpacity onLongPress={onLongPress} onPress={undefined} style={{marginHorizontal: 16}}>
      <RText
        // numberOfLines={2}
        // must wait for solution to https://github.com/facebook/react-native/issues/14981
        // adjustsFontSizeToFit
        // minimumFontScale={0.8}
        size={16}
        color={colors.DARK_PURPLE}
        style={[titleStyle, {textAlign: 'center'}]}
        numberOfLines={1}
      >
        {bot.error ? 'Unavailable' : bot.title}
      </RText>
      <RText
        // minimumFontScale={0.6}
        numberOfLines={1}
        weight="Light"
        size={12}
        color={colors.DARK_PURPLE}
        style={{textAlign: 'center'}}
      >
        {bot.addressData && bot.addressData.locationShort}
      </RText>
    </TouchableOpacity>
  )
}

const Loader = () => (
  <View
    style={{
      alignItems: 'center',
      paddingTop: 20 * k,
      paddingBottom: 80 * k,
      backgroundColor: 'white',
    }}
  >
    <Spinner />
  </View>
)
