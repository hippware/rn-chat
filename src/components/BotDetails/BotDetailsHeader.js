// @flow

import React from 'react';
import {View, Text, Animated, Alert, TouchableWithoutFeedback, Image, StyleSheet} from 'react-native';
// import {observable, toJS} from 'mobx';
import Popover from 'react-native-popover';
import {observer} from 'mobx-react/native';
// import Screen from '../Screen';
import botFactory from '../../factory/botFactory';
import {k, width, defaultCover} from '../Global';
import botStore from '../../store/botStore';
import locationStore from '../../store/locationStore';
import {colors} from '../../constants';
import BotButtons from './BotButtons';
import UserInfoRow from './UserInfoRow';
import Bot from '../../model/Bot';
// import BotPostCard from './BotPostCard';
// import ListFooter from '../ListFooter';
// import AddBotPost from './AddBotPost';
import {RText} from '../common';

type Props = {
  botId: string,
  flashPopover: Function,
};

type State = {
  fadeAnim: any,
  // showNavBar: true,
  // navBarHeight: any,
  // currentScreenWidth?: number,
  // currentScreenHeight?: number,
  isVisible?: boolean,
  buttonRect?: Object,
};

const DOUBLE_PRESS_DELAY = 300;

class BotDetailsHeader extends React.Component {
  props: Props;
  state: State;
  bot: ?Bot;
  lastImagePress: ?number;
  userInfo: any;

  constructor(props: Props) {
    super(props);
    this.state = {
      isVisible: false,
      fadeAnim: new Animated.Value(0),
      buttonRect: {},
    };
  }

  componentWillMount() {
    this.bot = botFactory.create({id: this.props.botId});
  }

  flashPopover = (buttonRect?: Object) => {
    this.setState({isVisible: true, buttonRect});
    setTimeout(() => this.setState({isVisible: false, buttonRect: {}}), 2000);
  };

  showPopover = () => {
    this.userInfo.measure()((ox, oy, w, h, px, py) => this.flashPopover({x: px, y: py, width: w, height: h}));
  };

  handleImagePress = (e: Object) => {
    const now = new Date().getTime();

    if (this.lastImagePress && now - this.lastImagePress < DOUBLE_PRESS_DELAY) {
      delete this.lastImagePress;
      this.handleImageDoublePress(e);
    } else {
      this.lastImagePress = now;
    }
  };

  handleImageDoublePress = () => {
    if (!this.bot.isSubscribed) {
      this.subscribe();
    }
  };

  unsubscribe = () => {
    Alert.alert(null, 'Are you sure you want to unsubscribe?', [
      {text: 'Cancel', style: 'cancel'},
      {
        text: 'Unsubscribe',
        style: 'destructive',
        onPress: () => botStore.unsubscribe(this.bot),
      },
    ]);
  };

  subscribe = () => {
    botStore.subscribe(this.bot);
    // do animation
    this.setState({fadeAnim: new Animated.Value(1)});
    setTimeout(() => {
      Animated.timing(this.state.fadeAnim, {toValue: 0}).start();
    }, 500);
  };

  render() {
    const {bot} = this;
    if (!bot) return null;
    const {owner} = this.bot;
    const isOwn = !owner || owner.isOwn;
    return (
      <View style={{flex: 1}}>
        <View style={{height: width, backgroundColor: 'white'}}>
          <TouchableWithoutFeedback onPress={this.handleImagePress}>
            {bot.image && bot.image.source ? (
              <Image style={{height: width, width}} resizeMode='contain' source={bot.image.source} />
            ) : (
              <Image style={{height: width, width}} source={defaultCover[bot.coverColor % 4]} resizeMode='contain' />
            )}
          </TouchableWithoutFeedback>
          <Animated.View pointerEvents='none' style={[{opacity: this.state.fadeAnim}, styles.botAddedContainer]}>
            <Image source={require('../../../images/iconBotAdded.png')} />
          </Animated.View>
        </View>
        <BotButtons isOwn={isOwn} bot={bot} subscribe={this.subscribe} unsubscribe={this.unsubscribe} isSubscribed={bot.isSubscribed} afterCopy={this.showPopover} />
        <UserInfoRow flashPopover={this.props.flashPopover} bot={bot} owner={owner} ref={r => (this.userInfo = r)} />
        {!!bot.description && (
          <View style={styles.descriptionContainer}>
            <RText numberOfLines={0} size={16} weight='Light' color={locationStore.isDay ? colors.DARK_PURPLE : colors.WHITE}>
              {bot.description}
            </RText>
          </View>
        )}
        <View style={{height: 8.5, width}} />
        <View style={{height: 45, width, flexDirection: 'row', alignItems: 'center', backgroundColor: 'white'}}>
          <Image style={{marginLeft: 14, width: 14, height: 14}} source={require('../../../images/postsIcon.png')} />
          <RText size={15} color={colors.DARK_PURPLE} style={{marginLeft: 7, letterSpacing: 0.3}}>
            Posts
          </RText>

          <RText size={12} color={colors.DARK_GREY} style={{marginLeft: 7}}>
            {bot.totalItems}
          </RText>
        </View>
        <View style={{height: 1, width}} />
        <Popover
          isVisible={this.state.isVisible}
          fromRect={this.state.buttonRect}
          contentStyle={{backgroundColor: colors.DARK_PURPLE}}
          placement='bottom'
          // onClose={this.closePopover}
        >
          <Text style={styles.popoverText}>Address copied to clipboard</Text>
        </Popover>
      </View>
    );
  }
}

export default observer(BotDetailsHeader);

const styles = StyleSheet.create({
  popoverText: {
    fontFamily: 'Roboto-Regular',
    color: 'white',
    fontSize: 14 * k,
  },
  showNoMore: {
    paddingTop: 10,
    alignItems: 'center',
    paddingBottom: 21,
  },
  noPhotosAdded: {
    fontFamily: 'Roboto-Regular',
    fontSize: 15,
    // @TODO: #683. standardize. this is between GREY and LIGHT_GREY
    color: 'rgb(186,186,186)',
  },
  attachPhoto: {
    height: 201 * k,
    backgroundColor: colors.LIGHT_GREY,
    justifyContent: 'center',
    alignItems: 'center',
  },
  descriptionContainer: {
    paddingLeft: 20 * k,
    paddingRight: 20 * k,
    paddingBottom: 15 * k,
    backgroundColor: colors.WHITE,
  },
  botAddedContainer: {
    height: width,
    position: 'absolute',
    left: 0,
    right: 0,
    justifyContent: 'center',
    alignItems: 'center',
  },
});
