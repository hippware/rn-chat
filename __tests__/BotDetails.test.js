// @flow
/* global test */

import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import BotDetails from '../src/components/BotDetails';
import BotDetailsHeader from '../src/components/BotDetails';
import {Provider} from 'mobx-react/native';
import {Bot} from 'wocky-client';
import injects from './utils/inject-mocks';

describe('BotDetails', () => {
  const wocky = {};

  test('renders with no data', () => {
    const toRender = (
      <Provider wocky={wocky} {...injects}>
        <BotDetails />
      </Provider>
    );
    const tree = renderer.create(toRender).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('renders with test bot', () => {
    const bot = Bot.create({
      id: '1234',
      isSubscribed: false,
      title: 'title',
      server: 'server',
      radius: 30,
      owner: null,
      // image: FileRef,
      // description: types.maybe(types.string),
      // visibility: VISIBILITY_PUBLIC,
      // location: types.maybe(Location),
      // address: '',
      // followersSize: 0,
      // totalItems: 0,
      // addressData: types.optional(Address, {}),
      // subscribers: types.optional(ProfilePaginableList, {}),
      // posts: types.optional(BotPostPaginableList, {}),
      // error: ''
    });
    const toRender = (
      <Provider wocky={wocky} {...injects}>
        <BotDetailsHeader bot={bot} scale={0.5} />
      </Provider>
    );
    const tree = renderer.create(toRender).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
