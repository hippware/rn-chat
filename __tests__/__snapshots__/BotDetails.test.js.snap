// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BotDetails renders with no data 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "white",
        "paddingBottom": 160,
        "paddingTop": 40,
      }
    }
  >
    <Image
      resizeMode="contain"
      source={
        Object {
          "testUri": "../../../images/loader_pink.gif",
        }
      }
      style={
        Array [
          undefined,
          Object {
            "height": 72,
            "width": 72,
          },
        ]
      }
    />
  </View>
</View>
`;

exports[`BotDetails renders with test bot 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "white",
        "paddingBottom": 160,
        "paddingTop": 40,
      }
    }
  >
    <Image
      resizeMode="contain"
      source={
        Object {
          "testUri": "../../../images/loader_pink.gif",
        }
      }
      style={
        Array [
          undefined,
          Object {
            "height": 72,
            "width": 72,
          },
        ]
      }
    />
  </View>
</View>
`;
